Task 3_1

O-Hlushchevska opened the pull request:
Add Factories for each type of class. 
Work with arrays + personal tasks

O-Hlushchevska committed:
Merge pull request #1 from O-Hlushchevska/Task-2-branch

Add destructor, population counter, food variation and daily portion calculation

O-Hlushchevska committed:
Merge branch 'Task-3' of https://github.com/O-Hlushchevska/TestApp into Task-3

O-Hlushchevska committed:
changes

O-Hlushchevska committed:
test check

O-Hlushchevska committed:
Changes in the arrays

Add Factories for each type of classes

MlikeW commented on TestApp/CatFactory.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class CatFactory
+    {
+        public static Pet Mara => new Pet("Cat", "Mara", DateTime.Parse("25/08/2016"), "Female", "British", 5.54);
```
Create enum for pet types
(e.g. public enum PetTypes{ Cat, Dog })

MlikeW commented on TestApp/CatFactory.cs:
```diff
@@ -0,0 +1,16 @@
+    public class CatFactory
+    {
+        public static Pet Mara => new Pet("Cat", "Mara", DateTime.Parse("25/08/2016"), "Female", "British", 5.54);
+        public static Pet Mammi => new Pet("Cat", "Mammi", DateTime.Parse("12/07/2020"), "Female", "British", 6.94);
```
+enum

MlikeW commented on TestApp/DogFactory.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class DogFactory
+    {
+        public static Pet Eva => new Pet("Dog", "Eva", DateTime.Parse("09/12/2015"), "Female", "doggy", 50.54);
```
one style name

MlikeW commented on TestApp/Pet.cs:
```diff
@@ -2,6 +2,8 @@
 {
     public class Pet
```
try to apply : IDisposable
and watch bahavior

MlikeW commented on TestApp/Pet.cs:
```diff
@@ -2,6 +2,8 @@
 {
     public class Pet
     {
+        public static bool shouldPrint = true;
```
make method non-static
make a parameter shouldPrint with default value true
public void print(bool isBold = false){ }

O-Hlushchevska replied:
The main idea of the static field `shouldPrint` is to control printing inside Pet's destructor, so, that I could later disable unnecessary lines in the console.

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,96 @@
+        private static void TestSingleArray() 
+        {
+            // array with size [4], fill values 0 and 1, try to get values from 3
+            int[] taskArray = new int[4];
```
try object[]

MlikeW replied:
try float[], long[], string[].....few types

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,96 @@
+            tempList.Insert(4, 9);
+            var arrayWIthAddedVal = tempList.ToArray();
+            Console.WriteLine($"\nCurrent state of the array: {string.Join(" ", arrayWIthAddedVal)}");
+            Console.WriteLine($"\nCurrent length of the array: {string.Join(" ", arrayWIthAddedVal.Length)}");
```
try to do using cycles

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,96 @@
+            Console.WriteLine($"\nCurrent length of the array: {string.Join(" ", arrayAfterDeletion.Length)}");
+
+            // Insert to [4] new element
+            var tempList = arrayAfterDeletion.ToList();
```
without usimg List

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,96 @@
+        private static void TestMultiArray()
+        {
+            // Multi-dimensional array
+            int[,] task2DArray = { { 1, 2 }, { 3, 4 }, { 5, 6 }, { 7, 8 } };
```
do same with Pets

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,96 @@
+            int[] taskDeleteInArray = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
+
+            // Permanently delete elements [5],[6],[7]
+            var arrayAfterDeletion = taskDeleteInArray.Where((val, idx) => idx != 5 && idx != 6 && idx != 7).ToArray();
```
Linq

O-Hlushchevska replied:
Did you expect something like that? 
var arrayAfterDeletion = (
    from idx in Enumerable.Range(0, taskDeleteInArray.Length)
    where idx != 5 && idx != 6 && idx != 7
    select taskDeleteInArray[idx]
).ToArray();

O-Hlushchevska committed:
Arrays' attraction :)

Add arrays with different types. 
Small loop.
Template for collections.

MlikeW commented on TestApp/Food.cs:
```diff
@@ -2,34 +2,33 @@
-        private char typeOfFood;
-        private DateTime expirationDate;
-        private double nutrition;
+        public enum TypeOfFood
```
create folder and put enums there as separate files

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,197 @@
+            string[] stringArray = new string[4];
+            stringArray[0] = "Mango";
+            stringArray[1] = "Banana";
+            Console.WriteLine($"\n3rd element of the string array is {stringArray[3]}");
```
+try struct, bool, char, enum + int?, bool? for a pssible with ?

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,197 @@
+
+
+            // Permanently delete elements [5],[6],[7]
+            var arrayAfterDeletionInt = taskDeleteInIntArray.Where((val, idx) => idx != 5 && idx != 6 && idx != 7).ToArray();
```
without linq (for, foreach, while) - for different types - different cycles

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,197 @@
+
+            int indexToInsert = 4;
+            int elementToInsert = 444;
+            Array.Resize(ref arrayAfterDeletionInt, arrayAfterDeletionInt.Length + 1);
```
read about 'ref'

MlikeW replied:
and about 'out'

MlikeW replied:
value vs reference types

MlikeW replied:
> value vs reference types

+define ref after that

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,197 @@
+            Console.WriteLine($"\nThe sum of all elements in jagged array is {sumOfJaggedArray}");
+
+            // Print elements in reverse order
+            var reverseOfJaggedArray = jaggedArray.SelectMany(item => item.Cast<int>()).Reverse();
```
formulate more accurate description of difference Select vs SelectMany

MlikeW commented on TestApp/Task3_2.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+
+        private static void TestConcurentCollection()
```
redundant

O-Hlushchevska committed:
Part of Collections + comments

Part of the loops usage
Actions with standard collections
Part of the generic usage
Enum moved to separate file

MlikeW commented on TestApp/Enums.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace TestApp
+{
+    public enum FoodType
```
bad file naming
fix
way to resolve - two enums - two files - can be in some separatew folder

MlikeW commented on TestApp/Enums.cs:
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
```
check whether usings are redundant

MlikeW commented on TestApp/Enums.cs:
```diff
@@ -0,0 +1,20 @@
+    }
+    public enum FoodSpeciality
+    {
+        Wet,
```
add another enum value
e.g. 
Wet = 1,
Dry = 2

and use enum types using those values

MlikeW commented on TestApp/Pet.cs:
```diff
@@ -1,19 +1,29 @@
-    public class Pet
+    public class Pet:IDisposable
     {
+        public enum PetType
```
to separate file

MlikeW commented on TestApp/Pet.cs:
```diff
@@ -22,13 +32,22 @@ public Pet(string type, string name, DateTime dateOfBirth, string sex, string br
+        //        Console.WriteLine($"Pet {Name} is gone. You have {population} pets left");
+        //}
+
+        public void Dispose() 
```
more practice

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+
+        private static void TestArray<T>(T[] array, Func<T[], int[], T[]> delete, Func<T[], T, int, T[]> insert, T elementToInsert)
+        {
+            // Permanently delete elements [5],[6],[7]
```
read about regions

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+
+        private static T[] DeleteUsingLinq<T>(T[] array, int[] at)
+        {
+            return array.Where((val, idx) => !at.Contains(idx)).ToArray();
```
read what is linq, for what and from what

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+            return array.Where((val, idx) => !at.Contains(idx)).ToArray();
+        }
+
+        private static T[] DeleteUsingForLoop<T>(T[] array, int[] at)
```
Why 'using'?

MlikeW replied:
at = index of elements to delete

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+            TestArray(taskDeleteInEnumArray, DeleteUsingForLoop, InsertUsingWhileLoop, FoodType.Kormik);
+        }
+
+        private static void TestArray<T>(T[] array, Func<T[], int[], T[]> delete, Func<T[], T, int, T[]> insert, T elementToInsert)
```
different static/non-stativ fields/methods/...

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+            return result;
+        }
+
+        private static T[] DeleteUsingForeachLoop<T>(T[] array, int[] at)
```
better naming (don;t understand purpose)

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+
+        private static T[] DeleteUsingForeachLoop<T>(T[] array, int[] at)
+        {
+            int i = 0;
```
same naming thing

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+            T[] result = new T[array.Length - at.Length];
+            foreach (var item in array)
+            {
+                if (at.Contains(i))
```
if foreach try to not operate with index
or find a method to get that index

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+        }
+
+
+        private static T[] DeleteUsingLinq<T>(T[] array, int[] at)
```
get rid of int[] at

MlikeW replied:
animals = [Cat, Dog, Parrot, Rabbit]
del Parrot
det(animals, [1,2,3,4]) - no
only det(animals)
for all cycles

MlikeW commented on TestApp/Task3_1.cs:
```diff
@@ -0,0 +1,346 @@
+            T[] result = new T[array.Length - at.Length];
+            int skipped = 0;
+            int i = 0;
+            while (i < array.Length)
```
while(true) read about
